8Knot/pages/codebase/visualizations/reviewer_file_heatmap.py

```diff
@@ -0,0 +1,484 @@
+
+
+def process_data(
+    df_file: pd.DataFrame,
+    df_actions: pd.DataFrame,
+    df_file_cntbs: pd.DataFrame,
+    directory,
+    bot_switch,
+):
+
+    # df_file preprocessing
+    df_file = df_file_clean(df_file, df_file_cntbs, bot_switch)
+
+    df_dynamic_directory = cntrb_per_directory_value(directory, df_file)
+
+    # work around for using functions, will clean later
+    if df_dynamic_directory.empty:
+        return df_dynamic_directory
+
+    df_dynamic_directory = cntrb_to_last_activity(df_actions, df_dynamic_directory)
+
+    final = file_cntrb_activity_by_month(df_dynamic_directory, df_actions)
+
+    return final
+
+
+def create_figure(df: pd.DataFrame):
+    fig = px.imshow(
+        df,
+        labels=dict(x="Time", y="Directory Entries", color="Contributors"),
+        color_continuous_scale=px.colors.sequential.deep,
+    )
+
+    fig["layout"]["yaxis"]["tickmode"] = "linear"
+    fig["layout"]["height"] = 700
+    fig["layout"]["coloraxis_colorbar_x"] = -0.15
+    fig["layout"]["yaxis"]["side"] = "right"
+
+    return fig
+
+
+def df_file_clean(df_file: pd.DataFrame, df_file_cntbs: pd.DataFrame, bot_switch):
+    # strings to hold the values for each column (always the same for every row of this query)
+    repo_name = df_file["repo_name"].iloc[0]
+    repo_path = df_file["repo_path"].iloc[0]
+    repo_id = str(df_file["repo_id"].iloc[0])
+
+    # pattern found in each file path, used to slice to get only the root file path
+    path_slice = repo_id + "-" + repo_path + "/" + repo_name + "/"
+    df_file["file_path"] = df_file["file_path"].str.rsplit(path_slice, n=1).str[1]
+
+    # drop unneccessary columns not needed after preprocessing steps
+    df_file = df_file.reset_index()
+    df_file.drop(["index", "repo_name", "repo_path", "rl_analysis_date"], axis=1, inplace=True)
+
+    # split file path by directory
+    df_file = df_file.join(df_file["file_path"].str.split("/", expand=True))
+
+    # drop unnecessary columns
+    df_file.drop(["repo_id"], axis=1, inplace=True)
+    df_file_cntbs.drop(["repo_id", "cntrb_ids"], axis=1, inplace=True)
+
+    # Left join on df_files to only get the files that are currently in the repository
+    # and the contributors that have ever reviewed a pr that included edits on the file
+    df_file = pd.merge(df_file, df_file_cntbs, on="file_path", how="left")
+    # replace nan with empty string to avoid errors in list comprehension
+    df_file.reviewer_ids.fillna("", inplace=True)
+
+    # reformat reviewer_ids to list and remove bots if filter is on
+    if bot_switch:
+        df_file["reviewer_ids"] = df_file.apply(
+            lambda row: [x for x in row.reviewer_ids if x not in app.bots_list],
+            axis=1,
+        )
+    else:
+        df_file["reviewer_ids"] = df_file.apply(
+            lambda row: [x for x in row.reviewer_ids],
+            axis=1,
+        )
+
+    return df_file
+
+
+def cntrb_per_directory_value(directory, df_file):
```

Comment:
cntrbs_per_directory_level maybe?

Reply:
count_cntrbs_per_member_of_directory_level is long-winded but clearer

Reply:
I think "directory level" is a bit misleading. This function is not getting all of the contributors at different directory levels, it is taking all of the files and sub-directories of a specific directory and getting the contributors for those. Def think there could be a better naming than what it currently is